hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -118,6 +118,15 @@ object HoodieSparkSqlWriter {
 
       operation = WriteOperationType.INSERT
     }
+    // If no record key field and precombine field is provided, assume it's an append-only workload and do bulk insert
+    if (!hoodieConfig.contains(RECORDKEY_FIELD) && !hoodieConfig.contains(PRECOMBINE_FIELD) &&
+      operation == WriteOperationType.UPSERT) {
```

Comment:
Also handle the other incremental write operations like INSERT and so on? we need to think through all Write operations

Reply:
i think if users intentionally set a particular operation type we should fail fast. Here, instead of checking for upsert operation type, we should be checking whether the user set no operation type.